pkg/reconciler/pipelinerun/pipelinerun_expired_test.go

```diff
@@ -0,0 +1,160 @@
+package pipelinerun
+
+import (
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline"
+	apispipeline "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/tektoncd/pipeline/test"
+	tb "github.com/tektoncd/pipeline/test/builder"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/apis"
+)
+
+func newPipelineRun(completionTime, failedTime apis.VolatileTime, ttl *metav1.Duration) *apispipeline.PipelineRun {
+	pr := tb.PipelineRun("test-pipeline-run-with-expiration-ttl", "foo",
+		tb.PipelineRunLabel("tekton.dev/pipeline", "test-pipeline"),
+		tb.PipelineRunLabel(pipeline.GroupName+pipeline.PipelineTaskLabelKey, "hello-world-1"),
+		tb.PipelineRunLabel("tekton.dev/pipelineRun", "test-pipeline-run-with-annotations"),
+		tb.PipelineRunAnnotation("PipelineRunAnnotation", "PipelineRunValue"),
+		tb.PipelineRunSpec(
+			"hello-world",
+			tb.PipelineRunServiceAccountName("test-sa"),
+		),
+	)
+
+	if !completionTime.Inner.IsZero() {
+		c := apis.Condition{Type: apis.ConditionSucceeded, Status: v1.ConditionTrue, LastTransitionTime: completionTime}
+		pr.Status.Conditions = append(pr.Status.Conditions, c)
+	}
+
+	if !failedTime.Inner.IsZero() {
+		c := apis.Condition{Type: apis.ConditionSucceeded, Status: v1.ConditionFalse, LastTransitionTime: failedTime}
+		pr.Status.Conditions = append(pr.Status.Conditions, c)
+	}
+
+	if ttl != nil {
+		pr.Spec.ExpirationSecondsTTL = ttl
+	}
+
+	return pr
+}
+
+func durationPointer(n int) *time.Duration {
+	s := time.Duration(n) * time.Second
+	return &s
+}
+
+func TestTimeLeft(t *testing.T) {
+	now := apis.VolatileTime{Inner: metav1.Now()}
+
+	PrTestCases := []struct {
+		name             string
+		completionTime   apis.VolatileTime
+		failedTime       apis.VolatileTime
+		ttl              *metav1.Duration
+		since            *time.Time
+		expectErr        bool
+		expectErrStr     string
+		expectedTimeLeft *time.Duration
+	}{
+		{
+			name:         "Error case: PipelineRun unfinished",
```

Comment:
For readability I think it might be a good idea to split this up into two different test funcs - the first can execute all the error cases and the second all the success cases.  This will reduce the number of `if` statements in the test and make the split between "this tests errors" and "this tests successes" clearer.

We then also wouldn't need the "Error case: " prefix in this name strings.